Add spam checking on pattern submissions

dd32 opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->

This PR adds some anti-spam checks to pattern submissions:
 - Akismet
 - Basic paragraph-only detection (ie. if a submission is only paragraphs, it's likely to require manual review / be an automated bot)
 - A keyword to use for testing the spam process.

In order to simplify the text being checked, I leveraged the Pattern Translator codebase to extract all strings from the submitted content, rather than simply passing the HTML to Akismet, to avoid any issues related to HTML.

<!-- Reference any related issues or PRs here. Each issue needs the "fixes" keyword if the PR fixes more than one thing. -->
See #22

<!-- Don't forget to update the title with something descriptive. -->

### How to test the changes in this Pull Request:

1. Apply PR
2. Create a Spam pattern. (Either by including `PatternDirectorySpamTest` within the content, or the pattern only containing `core/paragraph` blocks)
3. Verify that the post is pending review.

You can also try running `yarn test:php` however I've been unable to run phpunit locally due to `wp-env` problems, so I'm not able to check the unit test additions actually work.

<!-- If you can, add the appropriate [Component] label(s). -->

dd32 committed:
Add a filter to push patterns to pending status (and keep them there) if they don't meet the requirements.

dd32 committed:
Merge the code with the existing pattern-validation code.

dd32 committed:
Add missing class property definition

dd32 committed:
Run the submitted patterns past Akismet.

dd32 committed:
Extract all URLs

dd32 committed:
Treat paragraph-only submissions as likely spam.

dd32 committed:
Add tests to confirm detected-as-spam.

dd32 committed:
Fix linting issues.

dd32 committed:
Remove extra whitespace.

ryelle commented on public_html/wp-content/plugins/pattern-directory/includes/pattern-validation.php:
```diff
@@ -200,5 +204,141 @@ function validate_status( $prepared_post, $request ) {
+		);
+
+		$akismet = \Akismet::rest_auto_check_comment( $akismet_payload );
+		if ( is_wp_error( $akismet ) ) {
+			$akismet = array( 'akismet_result' => 'discard' );
+		}
```
Is this saying treat it as spam if Akismet errors?

dd32 replied:
No.

This is some uglyness from the Akismet plugin API not being designed to be used outside of comment filters.

If the "Discard the worst of the worst spam" is enabled, Akismet will return a WP_Error instead of changing the status to spam when it thinks it's being called from the REST API (that's why `rest_auto_check_comment` is called). If it thinks it's being called outside of the rest-api then it redirects and/or die's.

ryelle commented on public_html/wp-content/plugins/pattern-directory/includes/pattern-validation.php:
```diff
@@ -200,5 +204,141 @@ function validate_status( $prepared_post, $request ) {
+
+	// If it's been detected as spam, flag it as pending-review.
+	if ( $is_spam ) {
+		$prepared_post->post_status = 'pending';
```
Do you think we could also attach a note using the Internal Notes plugin? Otherwise it's not clear if this was pending on purpose (the default status is `pending`) vs spammed into pending.

dd32 replied:
Ah, I had thought the plan was for auto-publish, not always going to pending.
Would it make more sense to add a different post_status for `spam` then?

Doesn't look like Internal Notes is in the local environment, so I didn't realise it was being used.

dd32 replied:
>  I haven't been able to trigger Akismet spamming a pattern

I think about half of the seemingly-spam submissions to w.org/patterns triggered it (I copy pasted them over to local as a test), based on those posts I then added the `only-paragraphs` check.

I find it's harder to trigger Akismet with content than one would expect.. Looking at a public blogs spam comments should also pull up a few strings it'll trigger on.

dd32 replied:
> Ah, I had thought the plan was for auto-publish, not always going to pending.

Clarification on that, I had assumed this due to everyone having `publish_pattern` capabilities from abb07a078e3041b14abef37ed1fa77f25f66618e.  Upon reviewing this, it seems like this method will prevent `publish`, so that's fine.

dd32 replied:
> Would it make more sense to add a different post_status for spam then?

I've explored that in: https://github.com/dd32/pattern-directory/compare/add/pattern-validation...dd32:add/spam-status

ryelle replied:
> Ah, I had thought the plan was for auto-publish, not always going to pending

It's an option, I don't know which we'll launch with. I think the plan is to eventually auto-publish, but be pending at launch to add a little bit of a manual quality check.

> Doesn't look like Internal Notes is in the local environment, so I didn't realise it was being used.

Oh oops, it should get added to the composer deps.

ryelle replied:
> I find it's harder to trigger Akismet with content than one would expect.. Looking at a public blogs spam comments should also pull up a few strings it'll trigger on.

Finally got it to trigger, though I had to change the `comment_author` to a spammy name. The note added looks good 👍🏻

ryelle replied:
> Would it make more sense to add a different post_status for spam then?

I like `add/spam-status`, I think that would make it much easier to separate out spam especially with everything pending at launch. Since you've already got it in a separate branch, would you rather make a separate PR for it once this lands, or combine it all?

dd32 replied:
> I like add/spam-status, I think that would make it much easier to separate out spam especially with everything pending at launch.

Sounds good, I'll merge it into this PR.

> Since you've already got it in a separate branch

It was a branch forked from this branch, so I could quickly merge it into this branch if we went ahead with it.

ryelle reviewed (commented):
The tests pass locally, and [on the PR itself](https://github.com/WordPress/pattern-directory/runs/5300816056?check_suite_focus=true). Manually tested as well. I haven't been able to trigger Akismet spamming a pattern, but the other two checks work as expected.

dd32 committed:
Add an internal note when a post is marked as pending due to anti-spam detection.

dd32 committed:
Add an explicit Spam status, to differentiate detected-as-spam posts from other pending-review posts.

dd32 committed:
Whitespace is always good.

dd32 committed:
Update tests

ryelle commented on public_html/wp-content/plugins/pattern-directory/includes/pattern-validation.php:
```diff
@@ -200,5 +204,158 @@ function validate_status( $prepared_post, $request ) {
+	// Combine strings for ease of use.
+	$combined_strings = implode( "\n", $strings );
+	$combined_links   = implode( "\n", $links );
+	$combined         = $combined_strings . "\n" . $combined_links;
```
I don't see `$combined` used, should this be sent to akismet instead of just `$combined_strings`?

dd32 replied:
Yeah, `$combined` isn't being used, nor is `$links` / `$combined_links`.

I had added those with the intention of using them, but never did.. 

I'll pull those out for now.

dd32 committed:
Remove `$combined_links` & `$combined` as they're not currently used.

This leaves `$links` in place even though it's not in use, as it's likely needed in the future.

dd32 committed:
Remove `$links`, it's not used and we can add it later if needed.
